nemo/collections/asr/models/wav2vec/wav2vec_model.py

```diff
@@ -0,0 +1,858 @@
+import logging
+import math
+import pickle
+from typing import Optional, Tuple, List
+
+import numpy as np
+import torch
+import torch.nn.functional as F
+from omegaconf import DictConfig, OmegaConf
+from pytorch_lightning import Trainer
+from torch import nn
+
+from nemo.collections.asr.data.audio_to_text import TarredAudioToCharDataset, AudioToCharDataset
+from nemo.collections.asr.losses.wav2vecloss import Wav2vecCriterion
+from nemo.collections.asr.models.wav2vec.modules.config import TransformerSentenceEncoderConfig, \
+    TransformerEncoderConfig, QuantizeConfig, ConvFeaturesConfig, Wav2VecEncoderModelConfig, ConvConfig, LossConfig
+from nemo.collections.asr.models.wav2vec.modules.gumbel_vector_quantizer import GumbelVectorQuantizer
+from nemo.collections.asr.models.wav2vec.modules.multihead_attention import MultiheadAttention
+from nemo.collections.asr.models.wav2vec.modules.norm import Fp32LayerNorm, Fp32GroupNorm
+from nemo.collections.asr.models.wav2vec.modules.utils import compute_mask_indices
+from nemo.collections.asr.parts.perturb import process_augmentations
+from nemo.core import ModelPT
+from nemo.core.classes.common import PretrainedModelInfo
+
+
+def buffered_arange(max):
+    if not hasattr(buffered_arange, "buf"):
+        buffered_arange.buf = torch.LongTensor()
+    if max > buffered_arange.buf.numel():
+        buffered_arange.buf.resize_(max)
+        torch.arange(max, out=buffered_arange.buf)
+    return buffered_arange.buf[:max]
+
+
+class GradMultiply(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, x, scale):
+        ctx.scale = scale
+        res = x.new(x)
+        return res
+
+    @staticmethod
+    def backward(ctx, grad):
+        return grad * ctx.scale, None
+
+
+class Wav2VecEncoderModel(ModelPT):
```

Comment:
@okuchaiev which class Wav2VecEncoderModel should inherit from?

Reply:
ModelPT seems to be the right choice here because this is something that is trained - so needs to be a LightningModule